libs/pilight/protocols/433.92/hideki.c

```diff
@@ -0,0 +1,512 @@
+
+		json_append_member(hideki->message, "battery", json_mknumber(battery_ok, 0));
+		json_append_member(hideki->message, "channel", json_mknumber(channel, 0));
+		json_append_member(hideki->message, "rc", json_mknumber(rc, 0));
+	}
+}
+
+static int checkValues(struct JsonNode *jvalues) {
+	struct JsonNode *jid = NULL;
+
+#ifdef HIDEKI_DEBUG
+	logprintf(LOG_DEBUG, "HIDEKI checkValues() called");
+#endif
+
+	if((jid = json_find_member(jvalues, "id"))) {
+		struct settings_t *snode = NULL;
+		struct JsonNode *jchild = NULL;
+		struct JsonNode *jchild1 = NULL;
+		double channel = -1.0, rc = -1.0, tmp = -1.0;
+
+		jchild = json_first_child(jid);
+		while(jchild) {
+			jchild1 = json_first_child(jchild);
+			while(jchild1) {
+				if(strcmp(jchild1->key, "channel") == 0) {
+					channel = jchild1->number_;
+				}
+				if(strcmp(jchild1->key, "rc") == 0) {
+					rc =  jchild1->number_;
+				}
+				jchild1 = jchild1->next;
+			}
+			jchild = jchild->next;
+		}
+
+		if(channel < 0.0) {
+			logprintf(LOG_WARNING, "HIDEKI checkValues(): no channel found in given configuration as child of id");
+			return -1;
+		}
+		if(channel > 255.0) {
+			logprintf(LOG_ERR, "HIDEKI checkValues(): invalid channel value found in configuration (max: 255, got: %.0lf)", channel);
+			return -1;
+		}
+
+		if(get_settings((uint8_t)channel) == NULL) {
+			if((snode = MALLOC(sizeof(struct settings_t))) == NULL) {
+				fprintf(stderr, "out of memory\n");
+				exit(EXIT_FAILURE);
+			}
+			snode->rc = rc;  // rolling code might be misused as key, iff its value is fix for a device
+			snode->calibration = CALIBRATION_COUNT;
+			snode->sensortype = HIDEKI_UNKNOWN;
+			snode->channel = channel;
+			snode->temp = 0;
+			snode->humi = 0;
+			snode->wind_dir = 0;
+			snode->wind_factor = 1.0;
+			snode->rain = 1.0;
+			snode->ignore_temp = 0;
+
+			json_find_number(jvalues, "temperature-offset", &snode->temp);
+			json_find_number(jvalues, "humidity-offset", &snode->humi);
+			json_find_number(jvalues, "wind-direction-offset", &snode->wind_dir);
+			json_find_number(jvalues, "wind-factor", &snode->wind_factor);
+			json_find_number(jvalues, "rain-factor", &snode->rain);
+			if(json_find_number(jvalues, "ignore-temperature-on-battery-low", &tmp) == 0) {
+				snode->ignore_temp = (uint8_t)round(tmp);
+			}
+			snode->next = settings;
+			settings = snode;
+#ifdef HIDEKI_DEBUG
+			logprintf(LOG_DEBUG, "HIDEKI checkValues(): new setting appended");
+#endif
+		}
+	}
+	return 0;
+}
+
+static void gc(void) {
+	struct settings_t *tmp = NULL;
+	while(settings) {
+		tmp = settings;
+		settings = settings->next;
+		FREE(tmp);
+	}
+	if(settings != NULL) {
+		FREE(settings);
+	}
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void hidekiInit(void) {
+	protocol_register(&hideki);
+	protocol_set_id(hideki, "hideki");
+	protocol_device_add(hideki, "TS04", "HIDEKI TS04 sensor");
+	protocol_device_add(hideki, "Hideki-Wind", "HIDEKI Wind sensor");
```

Comment:
Above you wrote that this is not something we can use inside a pilight protocol. Why still define it here?

Reply:
if you refer to the code below the statement "if(parity != byteParity(packet[i]))" then I want to note that the sensortype variable is initialized with HIDEKI_WIND. This value stays valid in case that for all iterations of the for-loop the check "parity != byteParity(packet[i]))" is false.

Reply:
But that piece of code isn't affected in any way by the `protocol_device_add` pieces. I would just remove those aliasses.

Reply:
Ok, that was not so clear to me. In that case I'm wondering why other protocol implementations also use protocol_device_add() multiple times, e.g. tfa.c has 5 calls of protocol_device_add(). I think, multiple aliases can enable the user to setup the corresponding string into the config file to improve the readability of the config file. From my point of view it would be nice to get access to these settings from inside the checkValues() function. In that case pilight could know the sensortype that the user had configured.

Reply:
I would like to know your opinion to the following issue: Since I did not identify a unique ID for the sensors, I basically use the channel number to distinguish the devices. My test devices can be configured only for 3 different channels. So, theoretically I can only have 3 sensors of that type. On the other hand, the sensores transmit a "rolling code", that could be used together with the channel number to distinguish the sensors. But there are disadvantages: The value of the rc changes on every battery change or reset of the sensor and the user has no influence to its value. After each battery change the rc canges to an unpredictable value and the user can only read the value by pilight-receive or something similar. So, a change of the config file would be necessary. What is your opinion: Does it make sense to include the rc value into the "id" section of the config file?

Reply:
> protocol_device_add() multiple times

Because the same protocol can be implemented in different brands. In the future i want to remove those aliasses and just call them 'switch1', 'switch2', 'switch3' etc. And the same for the sensors, weather, dimmers etc.

> I think, multiple aliases can enable the user to setup the corresponding string into the config file to improve the readability of the config file

True

> In that case pilight could know the sensortype that the user had configured.

Why would that be usefull. Why not just have one configured device that stores all those values?

> Does it make sense to include the rc value into the "id" section of the config file?

Depends how much the batteries needs changed? If that's once every let's say 6 months, i would include it. If it's once each month, i would exclude it. One idea that comes up right now is the option of allowing a wild card. So that the device reacts to each channel or id.

Reply:
> Why would that be usefull. Why not just have one configured device that stores all those values?

Because in my case the sensor type can be an additional criteria (id) to identify the device. So, devices of different types can share the same channel. 
> One idea that comes up right now is the option of allowing a wild card

This is a good idea. I will think about it: maybe the rolling code can be included into the id section and will only be used as criteria if the configured value is >=0

Reply:
> Because in my case the sensor type can be an additional criteria (id) to identify the device. So, devices of different types can share the same channel.

Are those wind and rain meters just the same hardware device?

> This is a good idea.

At the moment however, this is still not possible.

Reply:
I have no idea. I have never tested wind or rain meters.

Reply:
Shall we just assume for now that those are the same devices and let the protocol communicate it that way?

Reply:
Ok.

Reply:
> 
> 
> I would like to know your opinion to the following issue: Since I did not identify a unique ID for the sensors, I basically use the channel number to distinguish the devices. My test devices can be configured only for 3 different channels. So, theoretically I can only have 3 sensors of that type. On the other hand, the sensores transmit a "rolling code", that could be used together with the channel number to distinguish the sensors. But there are disadvantages: The value of the rc changes on every battery change or reset of the sensor and the user has no influence to its value. After each battery change the rc canges to an unpredictable value and the user can only read the value by pilight-receive or something similar. So, a change of the config file would be necessary. What is your opinion: Does it make sense to include the rc value into the "id" section of the config file?